File: word-search.py
from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])

        def match(r: int, c: int, i: int) -> bool:
            if i == len(word):
                return True
            if r not in range(rows) or c not in range(cols):
                return False
            
            ch = board[r][c]
            board[r][c] = ""
            if (word[i] == ch
                and (
                    match(r+1, c, i+1)
                    or match(r, c+1, i+1)
                    or match(r-1, c, i+1)
                    or match(r, c-1, i+1)
                )
            ):
                return True
            board[r][c] = ch
            return False
        
        for r in range(rows):
            for c  in range(cols):
                if word[0] == board[r][c] and match(r, c, 0):
                    return True
        return False

File: word-search.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        for (int r=0; r<board.size(); ++r) {
			for (int c=0; c<board[0].size(); ++c) {
				if (board[r][c] == word[0] && matches(r, c, 0, board, word)) {
					return true;
				}
			}
		}
		return false;
    }

	bool matches(int r, int c, int i, vector<vector<char>>& board, string& word) {
		if (i == word.size()) return true;
		if ((r < 0) || (r >= board.size()) || (c < 0) || (c >= board[0].size())) return false;

		char ch = board[r][c];
		board[r][c] = '\0';
		
		if (
			(ch == word[i])
			&& (
				matches(r+1, c, i+1, board, word)
				|| matches(r, c+1, i+1, board, word)
				|| matches(r-1, c, i+1, board, word)
				|| matches(r, c-1, i+1, board, word)
			)
		) return true;

		board[r][c] = ch;
		return false;
	}
};
